src/Stratis.Bitcoin.Features.BlockStore/BlockStore.cs

```diff
@@ -0,0 +1,363 @@
+{
+    /// <summary>
+    /// Saves blocks to the database in batches, removes reorged blocks from the database.
+    /// </summary>
+    /// <remarks>
+    /// The batch is saved when total serialized size of all blocks in a batch reaches <see cref="BatchThresholdSizeBytes"/>,
+    /// when more than <see cref="BatchMaxSaveIntervalSeconds"/> passed since last batch was saved or when node is shutting down.
+    /// </remarks>
+    public class BlockStore : IDisposable
+    {
+        /// <summary>Maximum interval between saving batches.</summary>
+        private const int BatchMaxSaveIntervalSeconds = 37;
+
+        /// <summary>Maximum number of bytes the batch can hold until the downloaded blocks are stored to the disk.</summary>
+        private const uint BatchThresholdSizeBytes = 5 * 1000 * 1000;
+
+        /// <summary>The current batch size in bytes.</summary>
+        private int currentBatchSizeBytes;
+
+        /// <summary>The highest stored block in the repository.</summary>
+        public ChainedHeader StoreTip { get; private set; }
+
+        /// <summary>Items count in the <see cref="blocksQueue"/>.</summary>
+        internal int BlocksQueueCount => this.blocksQueue.Count;
+
+        /// <inheritdoc cref="ILogger"/>
+        private readonly ILogger logger;
+
+        /// <inheritdoc cref="INodeLifetime"/>
+        private readonly INodeLifetime nodeLifetime;
+
+        /// <inheritdoc cref="IChainState"/>
+        private readonly IChainState chainState;
+
+        /// <inheritdoc cref="StoreSettings"/>
+        private readonly StoreSettings storeSettings;
+
+        /// <inheritdoc cref="ConcurrentChain"/>
+        private readonly ConcurrentChain chain;
+
+        /// <inheritdoc cref="IBlockRepository"/>
+        private readonly IBlockRepository blockRepository;
+
+        /// <summary>Queue which contains blocks that should be saved to the database.</summary>
+        private readonly AsyncQueue<BlockPair> blocksQueue;
+
+        /// <summary>Task that runs <see cref="DequeueBlocksContinuouslyAsync"/>.</summary>
+        private Task dequeueLoopTask;
+
+        public BlockStore(
+            IBlockRepository blockRepository,
+            ConcurrentChain chain,
+            IChainState chainState,
+            StoreSettings storeSettings,
+            INodeLifetime nodeLifetime,
+            ILoggerFactory loggerFactory)
+        {
+            this.chainState = chainState;
+            this.nodeLifetime = nodeLifetime;
+            this.storeSettings = storeSettings;
+            this.chain = chain;
+            this.blockRepository = blockRepository;
+
+            this.blocksQueue = new AsyncQueue<BlockPair>();
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        /// <summary>
+        /// Initializes the <see cref="BlockStore"/>.
+        /// <para>
+        /// If StoreTip is <c>null</c>, the store is out of sync. This can happen when:</para>
```

Comment:
the logic inside initialise that does this, must be worthy of at least its own method expressing that?

Reply:
Yeah. Method called `RecoverStoreTipAsync` that handles exactly that situation

```
if (this.StoreTip == null)
    await this.RecoverStoreTipAsync().ConfigureAwait(false);
```

Reply:
So that method seems like it that should be the place for the detailed logic summary rather than here then ?

Reply:
It is. 

Component summary- overall overview of the whole component and it's high level logic. 
Method summary- overview of what this method does.